### PR TITLE
Added carbontracker script

### PR DESCRIPTION
**Pull request by ClaireDons**
Created carbon tracker workflow script and new workflow notebook.

### CODE REVIEW

File: src/excited_workflow/carbon_tracker.py [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+    Args:
+        df: dataframe which needs to be converted for training.
+        x_keys: list of input variables.
+        y_key: target variable name.
+
+    Returns:
+        Dataframe for training.
+    """
+    df_pycaret = df[x_keys + [y_key]]
+    df_reduced = df_pycaret[::10]
```
Comment (BSchilperoort) [inline src/excited_workflow/carbon_tracker.py:107]: We want to leave this out now this workflow is basically finished. It was mostly used to speed up testing.

File: src/excited_workflow/carbon_tracker.py [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+
+    return df_reduced
+
+
+def train_model(
```
Comment (BSchilperoort) [inline src/excited_workflow/carbon_tracker.py:112]: I think it's better to name this function differently, something like `groupwise_cross_validation`. Because the model training here is just to perform CV.

File: src/excited_workflow/carbon_tracker.py [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+    time = datetime.datetime.now().strftime("%Y-%m-%d_%H")
+    output_dir = output_path / f"carbon_tracker-{time}"
+    output_dir.mkdir(parents=True, exist_ok=True)
+
+    df_group = create_groups(ds, groups)
+
+    for group in range(groups):
+        target_ds, prediction = train_model(df_group, group, x_keys, y_key)
+        rmse = calculate_rmse(prediction["prediction_label"], target_ds[y_key])
+        rmse.to_netcdf(output_dir / f"rmse{group}.nc")
```
Comment (BSchilperoort) [inline src/excited_workflow/carbon_tracker.py:196]: can you also plot the rmse and write it as an image?

File: src/excited_workflow/carbon_tracker.py [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+        create_scatterplot(prediction["prediction_label"], target_ds[y_key])
+        plt.savefig(output_dir / f"scatter{group}.png")
+        plt.close()
+
+
+def save_model(
```
Comment (BSchilperoort) [inline src/excited_workflow/carbon_tracker.py:202]: Splitting this function up in training and saving is better I think

Comment (BSchilperoort) [reply src/excited_workflow/carbon_tracker.py:202]: It could be nice to make a generic train function, with an optional argument "group". This way the training function can be reused for the CV and the final model training.

**Review by BSchilperoort — COMMENTED** [review]
Hi Claire, nice work! 

I think only some names could be changed and functions refactored slightly, otherwise it's basically ready

File: carbon_tracker_workflow.ipynb [hunk 5]
```diff
@@ -0,0 +1,2510 @@
+{
```
Comment (BSchilperoort) [inline carbon_tracker_workflow.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    ds_cb = Path("/data/volume_2/EXCITED_prepped_data/CT2022.flux1x1-monthly.nc")
</pre><p>When a variable is named <code>ds_</code> I assume it's an xarray dataset. cb_file would be more clear I think. Or change the functions so that they take an xarray dataset instead of the path.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/EXCITED-CO2/excited-workflow/pull/36/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (BSchilperoort) [inline carbon_tracker_workflow.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    ds_input = carbon_tracker.merge_datasets(desired_data, ds_cb)
</pre><p>This now does loading and merging... Perhaps it's not needed to wrap that in a function, as it's already quite simple to load and marge.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/EXCITED-CO2/excited-workflow/pull/36/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (BSchilperoort) [inline carbon_tracker_workflow.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    ds_na = carbon_tracker.mask_region(ds_regions, ds_cb, ds_input)
</pre><p>Let's move the .sel operation which is in this function to the notebook. Or perhaps we can leave it out altogether by now (except for the time slice).</p><p>That would make this notebook able to do a worldwide model training (once we have global era5 data...)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/EXCITED-CO2/excited-workflow/pull/36/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (BSchilperoort) [inline carbon_tracker_workflow.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    carbon_tracker.validate_model(ds_na, 5, x_keys, y_key, output_path)
</pre><p>It's a shame those pycaret plots don't use scientific notation. Can you set round=7 or something to still show RMSE etc.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/EXCITED-CO2/excited-workflow/pull/36/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (BSchilperoort) [inline carbon_tracker_workflow.ipynb:1]: <p>The values aren't converted anymore</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/EXCITED-CO2/excited-workflow/pull/36/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (ClaireDons) [reply carbon_tracker_workflow.ipynb:1]: round=7 works but now I wonder if there is a better way to do it, is ok for now though.

Comment (ClaireDons) [reply carbon_tracker_workflow.ipynb:1]: I left it out for now, it seems to still work ok, but let me know if that is now what you meant.

File: src/excited_workflow/carbon_tracker.py (hunk 4 above)
Comment (ClaireDons) [reply src/excited_workflow/carbon_tracker.py:221]: Added a training function but I am unsure it works in the way you meant.

File: src/excited_workflow/carbon_tracker.py [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+    ds_cb = excited_workflow.utils.convert_timestamps(ds_cb)
+    ds_input = xr.merge(
+        [
+            datasets[name].load(freq="monthly", target_grid=ds_cb)
+            for name in desired_data
+        ]
+    )
+    ds_na = mask_region(regions_path, cb_file, "transcom_regions", ds_input)
+    validate_model(ds_na, 5, x_keys, y_key, output_dir)
+    save_model(ds_na, x_keys, y_key, output_dir)
```
Comment (BSchilperoort) [inline src/excited_workflow/carbon_tracker.py:275]:
```suggestion
    model = train_model(ds_na, x_keys, y_key)
    save_model(model, x_keys, y_key, output_dir)
```
I still think it would be nicer to split this up this way.

File: src/excited_workflow/carbon_tracker.py [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+def create_scatterplot(prediction: xr.DataArray, target: xr.DataArray) -> None:
+    """Create scatterplot of prediction vs. target.
+
+    Args:
+        prediction: dataframe column of prediction.
+        target: dataframe column of target variable.
+    """
+    plt.figure(figsize=(5, 5))
+    plt.scatter(prediction, target, s=25)
+    plt.xlabel("Prediction")
```
Comment (BSchilperoort) [inline src/excited_workflow/carbon_tracker.py:155]: Can you add a 1:1 line to this plot to make analysis easier?

File: src/excited_workflow/carbon_tracker.py [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+        ds: dataset for training.
+        groups: number of groups.
+        x_keys: list of input variables.
+        y_key: target variable name.
+        output_dir: directory to output rmse and scatterplots.
+    """
+    df_group = create_groups(ds, groups)
+
+    model_vars = [
+        f"{var}: {ds[var].attrs['long_name']}" if "long_name" in ds[var].attrs else var
```
Comment (BSchilperoort) [inline src/excited_workflow/carbon_tracker.py:175]:
```suggestion
        f"**{var}**: {ds[var].attrs['long_name']}" if "long_name" in ds[var].attrs else var
```
Might make the list more readable.

File: src/excited_workflow/carbon_tracker.py [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+        for var in x_keys
+    ]
+
+    text = "## Carbon tracker model \n \n ### Model variables: \n"
```
Comment (BSchilperoort) [inline src/excited_workflow/carbon_tracker.py:179]:
```suggestion
    text = "## Carbon tracker model \n \n### Model variables: \n\n"
```
I think in Markdown you're supposed to have an empty line before a table, otherwise things can (sometimes) break on some renderers. Safer to add it in.

File: src/excited_workflow/carbon_tracker.py [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+        rmse.to_netcdf(output_dir / f"rmse{group}.nc")
+        create_rmseplot(rmse)
+        plt.savefig(output_dir / f"rmseplot{group}.png")
+        plt.close()
+        create_scatterplot(prediction["prediction_label"], target_ds[y_key])
+        plt.savefig(output_dir / f"scatter{group}.png")
+        plt.close()
+        text = (
+            text
+            + f"\n ### Validation plots for group {group} \n"
```
Comment (BSchilperoort) [inline src/excited_workflow/carbon_tracker.py:198]:
```suggestion
            + f"\n### Validation plots for group {group} \n"
```
unnecessary whitespace

File: src/excited_workflow/carbon_tracker.py [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+    Args:
+        ds: dataset for training.
+        groups: number of groups.
+        x_keys: list of input variables.
+        y_key: target variable name.
+        output_dir: directory to output rmse and scatterplots.
+    """
+    df_group = create_groups(ds, groups)
+
+    model_vars = [
```
Comment (BSchilperoort) [inline src/excited_workflow/carbon_tracker.py:174]: Hmm. It might be nice to add the units here too. They should be available in most datasets.
For readability perhaps we split up the three in steps:

```py
long_names = [ds[var].attrs["long_name"] if "long_name" in ds[var].attrs else "undefined" for var in x_keys]
units = [ds[var].attrs["units"] if "units" in ds[var].attrs else "?" for var in x_keys]

for var, long_name, unit in zip(x_keys, long_names, units):
    text += f"- {var}: {long_name} ({unit})\n"

Comment (BSchilperoort) [reply src/excited_workflow/carbon_tracker.py:174]: Also note the `+=` for appending to a string

File: src/excited_workflow/carbon_tracker.py (hunk 9 above)
Comment (BSchilperoort) [reply src/excited_workflow/carbon_tracker.py:179]: Also, "Model variables (in order):" might be good to add. As it's the same order as required for the onnx file

File: src/excited_workflow/carbon_tracker.py [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+        f"{var}: {ds[var].attrs['long_name']}" if "long_name" in ds[var].attrs else var
+        for var in x_keys
+    ]
+
+    text = "## Carbon tracker model \n \n ### Model variables: \n"
+
+    for item in model_vars:
+        text = text + f"- {item} \n"
+
```
Comment (BSchilperoort) [inline src/excited_workflow/carbon_tracker.py:183]: Let's add the target variable name, long name and unit here as well.

File: workflow_carbon_tracker.ipynb
Comment (BSchilperoort) [inline workflow_carbon_tracker.ipynb:1]: Can you rename this file to `exploration_carbontracker_model.ipynb`?
Then you can also rename the new notebook to "workflow_carbontracker.ipynb`

File: src/excited_workflow/carbon_tracker.py [hunk 13]
```diff
@@ -0,0 +1,275 @@
...
+    for group in range(groups):
```
Comment (BSchilperoort) [inline src/excited_workflow/carbon_tracker.py:184]: You could change this step so all targets & predictions are in memory at once. E.g.:

```
rmses = []
for group in range(groups):
    ...
    rmses.append(calculate_rmse(...))

    for rmse in rmses:
        # do individual plots

    make_full_plot(rmses)
```

Comment (ClaireDons) [reply src/excited_workflow/carbon_tracker.py:294]: I ended up refactoring the validate model function a little more, which I hope is ok.

File: src/excited_workflow/carbon_tracker.py [hunk 14]
```diff
@@ -0,0 +1,371 @@
...
+        ]
+    )
+
+    ds_sel = ds_merged.sel({"time": slice("2000-01", "2019-12")})
+    ds_sel = ds_sel.compute()
+    ds_sel = ds_sel.where(ds_merged["transcom_regions"] == mask)
+    ds_sel[y_key].attrs["long_name"] = "terrestrial biosphere CO2 flux"
+    df_sel = ds_sel.to_dataframe().dropna()
+
+    return ds_sel, df_sel
```
Comment (BSchilperoort) [inline src/excited_workflow/carbon_tracker.py:56]: I'm not a big fan of returning the same data (essentially) twice... Let's leave it for now, I see why you did it. It just feels a bit clunky.

File: src/excited_workflow/carbon_tracker.py [hunk 15]
```diff
@@ -0,0 +1,371 @@
...
+        x_keys: list of input variables.
+        y_key: target variable name.
+        output_dir: directory to output rmse and scatterplots.
+    """
+    long_names = [
+        ds[var].attrs["long_name"] if "long_name" in ds[var].attrs else "undefined"
+        for var in x_keys
+    ]
+    units = [
+        ds[var].attrs["units"] if "units" in ds[var].attrs else "?" for var in x_keys
```
Comment (BSchilperoort) [inline src/excited_workflow/carbon_tracker.py:252]: image
I think we have to escape those `**` characters. They can be used as markdown formatting.

You could put the unit in between two ` characters.
```suggestion
        f"`{ds[var].attrs["units"]}`" if "units" in ds[var].attrs else "?" for var in x_keys
```

File: src/excited_workflow/carbon_tracker.py [hunk 16]
```diff
@@ -0,0 +1,371 @@
...
+    ]
+
+    text = "## Carbon tracker model \n \n### Model variables (in order): \n\n"
+
+    for var, long_name, unit in zip(x_keys, long_names, units, strict=True):
+        text += f"- {var}: {long_name} ({unit})\n"
+
+    text += "\n\n###Target variable:\n\n"
```
Comment (BSchilperoort) [inline src/excited_workflow/carbon_tracker.py:260]:
```suggestion
    text += "\n\n### Target variable:\n\n"
```
Otherwise the header doesn't work.

File: src/excited_workflow/carbon_tracker.py [hunk 17]
```diff
@@ -0,0 +1,371 @@
...
+        for var in x_keys
+    ]
+    units = [
+        ds[var].attrs["units"] if "units" in ds[var].attrs else "?" for var in x_keys
+    ]
+
+    text = "## Carbon tracker model \n \n### Model variables (in order): \n\n"
+
+    for var, long_name, unit in zip(x_keys, long_names, units, strict=True):
+        text += f"- {var}: {long_name} ({unit})\n"
```
Comment (BSchilperoort) [inline src/excited_workflow/carbon_tracker.py:258]:
```suggestion
        text += f"- **{var}**: {long_name} ({unit})\n"
```

File: src/excited_workflow/carbon_tracker.py [hunk 18]
```diff
@@ -0,0 +1,371 @@
...
+        rmse.plot()
+        plt.tight_layout()
+        plt.savefig(output_dir / f"rmseplot{idx}.png")
+        plt.close()
+        text += f"**RMSE map {idx}**\n ![image]({output_dir}/rmseplot{idx}.png)\n\n"
+
+    return text
+
+
+def create_markdown_file(
```
Comment (BSchilperoort) [inline src/excited_workflow/carbon_tracker.py:227]: I think it could be nice to add some explanatory text in the markdown file for people who aren't fully familiar with the workflow.

File: src/excited_workflow/carbon_tracker.py [hunk 19]
```diff
@@ -0,0 +1,371 @@
...
+        text += f"- {var}: {long_name} ({unit})\n"
+
+    text += "\n\n###Target variable:\n\n"
+    target_name = ds[y_key].attrs["long_name"]
+    target_units = ds[y_key].attrs["units"]
+    text += f"{y_key}: {target_name} ({target_units})\n"
+
+    text += "\n### Validation plots \n\n"
+
+    text = create_scatterplots(
```
Comment (BSchilperoort) [inline src/excited_workflow/carbon_tracker.py:267]:
```suggestion
    text += create_scatterplots(
```
Could you change these so the create_ functions return a string, which we can append here to the `text` variable?

**Review by BSchilperoort — APPROVED** [review]
Hi Claire, some final small comments from me.
Once you've resolved them feel free to merge (and bypass branch protection) ;)

File: src/excited_workflow/carbon_tracker.py (hunk 14 above)
Comment (ClaireDons) [reply src/excited_workflow/carbon_tracker.py:56]: Yea I felt the same and tried to find another way but couldn't think of anything better for now.
